Prevent repeated coercion of fields

metamben opened the pull request:
Fixes #22519.

It seems to me that the problem is that we don't distinguish fields coming from the source table from the references to those fields in outer queries based on the results of the selection. A similar issue arises in the SQL query processor when it nests queries containing expressions. There the problem is solved by marking the fields in the outer select as such and excluding the fields so marked from casting.

My solutions is attempt to implement this scheme for all situations when a field is a reference to a field returned from a nested query.

There are some specific points too that I wanted to highlight:

- The current implementation marks the fields after implicit fields have been added but before joins are (pre)processed.
- A namespaced keyword is used as a marker (a name is probably not the best one).
- Only the SQL driver honors the new marker (though the regression test seems to pass for all drivers supporting nesting).

*TODO*

- [x] Fix coerced fields "inherited" from joins.
- [x] Check DB types other than SQL.
  - It seems to me that other DB types don't support models at all... At least I couldn't create the model described in the bug ticket with a Mongo database.
- [x] Don't remove `:nested/outer` from the metadata, adjust the tests instead.

metamben committed:
Prevent duplicate coercion of fields

Fixes #22519.

metamben committed:
Add test reproducing the issue

metamben commented on src/metabase/driver/sql/query_processor.clj:
```diff
@@ -353,14 +353,15 @@
           source-alias         (field-source-alias field-clause)
           field                (when (integer? id-or-name)
                                  (qp.store/field id-or-name))
+          outer-select         (or (:metabase.query-processor.middleware.mark-outer-select-fields/outer-select options)
```
I don't like this name. Since it's not SQL specific, `select` doesn't sound well. (I will `:as-alias` the namespace, I think, when I get the name right.)

dpsutton replied:
Should we unify the keywords? Might as well replace all usages of `::nest-query/outer-select` with the new `mark-outer-select-fields/outer-select`? Or is that annoying for now or planned follow up work?

metamben replied:
If we agree that the meaning is the same, then we should use the same key, I think.

dpsutton replied:
Yes I believe these mean the same thing: a conversion has been applied already so don't convert again. We can also use a non-namespace aliased key `:nested/outer` or `:nested/inner` so we don't need to worry about requiring namespaces for the namespace of a key. 

If I'm misreading this then I apologize.

metamben commented on src/metabase/query_processor/middleware/annotate.clj:
```diff
@@ -193,7 +193,8 @@
                                                    (filter (fn [[k _v]]
-                                                             (and (keyword? k) (namespace k))))
+                                                             (and (keyword? k) (namespace k)
+                                                                  (not= k :metabase.query-processor.middleware.mark-outer-select-fields/outer-select))))
```
I'm not quite sure what the logic is behind `namespaced-options`. I'm removing this "annotation" because it's really only needed during compilation and leaving it here makes a lot of tests fail.

dpsutton replied:
I think we should check if we can get the tests passing. Perhaps we can do some judicious use of `partial=` to handle new keys? As it stands, this code makes me think its "special" rather than just unnecessary and removed for tests. Do you remember how many tests were failing?

dpsutton replied:
I would expect them to be different in exactly the current circumstance. Suppose `foo` is an integer column that has a `:Coercion/UNIXSeconds->DateTime` coercion on it. I would expect the query to be

`select foo from (select from_unixtime(foo) from t)`

I think we have to do the coercion at its lowest level. We really want no part of the query to see the integer of `foo`, it should be a timestamp type value throughout.

metamben replied:
I think you misunderstood what I meant. In your example what I wanted to say was that I expected the metadata to be the same for `select from_unixtime(foo) as foo from t` and `select foo from (select from_unixtime(foo) as foo from t)`. In other words, just wrapping a select `s` into a `select * from (s)` should not "change" the metadata.

metamben commented on src/metabase/query_processor/middleware/mark_outer_select_fields.clj:
```diff
@@ -0,0 +1,33 @@
+
+(defn- matching-fields?
+  [field source-field]
+  (every? #(= (field %) (source-field %)) [0 1]))
```
Is matching the type and the "ID" of the clause good (enough)?

dpsutton replied:
It's really hard to parse this. I'm assuming `field` is a map representing a field. You are treating it like a vector, maybe a field_ref?

metamben replied:
My expectation was that since both `field` and `source-field` are direct children of a `:fields` clause they have the form `[:field name-or-id opts]` and I check if the first two respective elements match. I was hoping that I can use simple equality, but the `opts` can be different because middlewares can add different annotations based on where the fields are in the query. I'm not sure though that `opts` can fully be ignored (i.e., if nothing there is part of the "identity" of the field).

dpsutton replied:
ah yeah they are field-refs then. I have `metabase.query-processor.util/field-ref->key` that I use. I've seen @camsaul has a similar function somewhere that grabs what's important out of the options. If you reuse the function i've listed we can more easily keep track of all of these and fix them at once.

metamben commented:
> we can attempt to do this where we "bubble up" the fields in the first place: `metabase.query-processor.middleware.annotate`: `flow-field-metadata` (i think that's the end of the list). Did you look at setting it from this side rather than trying to recognize it after the fact?

I was thinking about this, but haven't found the place. I'm going to have a look at it now.

metamben commented:
> `metabase.query-processor.middleware.annotate`: `flow-field-metadata`

@dpsutton Do you mean the call via the `add-column-info` middleware? It seems to me that's a post-processing middleware and runs after the compilation (and execution). If you mean something else, then what?

dpsutton commented:
dang. i think you are right. I'd have to spend more time looking for that seam. I'm sorry to have mislead you.

metamben committed:
Add test with specific field in the outer query

metamben committed:
Use library functions in implementing the middleware

metamben committed:
Unify ::nest-query/outer-select and ::outer/outer-select

The new common key is called :nested/outer.

metamben committed:
Get rid of SQL terminology in middleware name

The word "select" does not make much sense for some DB types.

metamben committed:
Unskip repro test

metamben committed:
Fix joins

metamben committed:
Do not remove the :nested/outer flag from the calculated metadata

metamben committed:
Fix e2e test

The turnIntoModel() model call already includes waiting for the dataset request so the second wait times out.

metamben commented:
@nemanjaglumac Are you OK with https://github.com/metabase/metabase/pull/24718/commits/a295fbcf51a0b76c0dbb9a3dab32cb72266114b5 fixing the repo test?

nemanjaglumac commented:
> @nemanjaglumac Are you OK with [a295fbc](https://github.com/metabase/metabase/commit/a295fbcf51a0b76c0dbb9a3dab32cb72266114b5) fixing the repo test?

@metamben absolutely. That was my mistake. Great catch! Thank you :)

dpsutton commented on src/metabase/query_processor/middleware/mark_outer_fields.clj:
```diff
@@ -0,0 +1,36 @@
+
+(defn- matching-fields?
+  [field source-field]
+  (= (qp.util/field-ref->key field) (qp.util/field-ref->key source-field)))
```
this expects a field_ref, not a field. You'll need `(comp qp.util/field-ref->key :field_ref)` as the function applied before checking equality.

metamben replied:
I'm confused. This is supposed to be called with two triples (vectors of length 3) with the first value being `:field`, the second an id or a name and we don't care about the third. Calling `:field_ref` would return `nil`. I'll have a look to see where `:field_ref`s come into play.

dpsutton commented:
This looks awesome. I think I'm seeing a scenario that it doesn't work though: when adding an order by clause.

```clojure
qp=> (mt/dataset sample-dataset
       (mt/with-temp-vals-in-db Field (mt/id :reviews :rating)
         {:coercion_strategy :Coercion/UNIXSeconds->DateTime
          :effective_type    :type/Instant}
         (mt/with-temp Card [{card-id :id}
                             {:dataset_query
                              (mt/mbql-query reviews
                                             {:fields [$rating]
                                              :limit  3})}]
           (qp/process-query {:type     :query
                              :database (mt/id)
                              :query    {:source-table (str "card__" card-id)
                                         :order-by [[(mt/id :reviews :rating) :desc]]}}))))
Execution error (NumberFormatException) at java.lang.NumberFormatException/forInputString (NumberFormatException.java:67).
For input string: "1970-01-01 00:00:05"
qp=> 
```

I was wondering what the `:nested/outer      true` is doing in `rewrite-fields-to-force-using-column-aliases`. It seems like some of the breakout fields and sorting goes through this but not all. So the above is running:

```sql
 SELECT   "source"."RATING" AS "RATING"
FROM     (
                SELECT timestampadd('second', "PUBLIC"."REVIEWS"."RATING", timestamp '1970-01-01T00:00:00Z') AS "RATING"
                FROM   "PUBLIC"."REVIEWS" limit 3) "source"
ORDER BY timestampadd('second', "source"."RATING", timestamp '1970-01-01T00:00:00Z') DESC limit 1048575 
```

So I wonder if its possible to have your walk descend into other clauses (breakouts, order bys) and add the metadata there. We want to cast once and only once correct?

metamben committed:
Handle breakout and ordering too

metamben commented:
Leaving `:nested/outer` in the metadata makes the models e2e test about 9 times out of 10.

metamben commented:
@dpsutton I've added support for breakout and ordering.

metamben commented on src/metabase/query_processor/middleware/mark_outer_fields.clj:
```diff
@@ -0,0 +1,32 @@
+                (seqable? source-fields)
+                (seq source-fields))
+         (let [id-set (into #{} (map second) source-fields)]
+           (-> form
+               (m/update-existing :fields annotate-outer-fields id-set)
+               (m/update-existing :breakout annotate-outer-fields id-set)
+               (m/update-existing :order-by annotate-outer-fields id-set)))
```
Could `reduce` if that's more readable.

dpsutton replied:
I think this is very readable as is.

dpsutton reviewed (approved):
Looks great to me. I think this is a nice improvement and should make queries more speedy.

metamben committed:
Merge branch 'master' into metamben/22519-prevent-duplcate-coercion

metamben committed:
Merge branch 'master' into metamben/22519-prevent-duplcate-coercion

metamben committed:
Merge branch 'master' into metamben/22519-prevent-duplcate-coercion
